### PR TITLE
PERFSCALE-1818: Initial commit of ansible-microshift automation

### PR DESCRIPTION
**Pull request by sjug**
- Pre-localhost setup and logging setup
- Add cadvisor to microshift node, run as service
- Fix systemd_service var and apply firewall rules sooner
- Manage prometheus config via template
- Need to restart cadvisor after CRI-O starts in order to pick up the CRI-O container metrics
- Split out EPEL repo install and add pbench agent tasks
- Add pbench to README & pbench_env to vars
- Set up pbench and kubectl
- Make pbench-user-benchmark async
- Updated inventory vars
- Move API checks to remote node and automate pbench data collection
- Write three stages of disk usage (pre-install, post-install, post-run) to file
- Modify timing tasks: reorder and small fixes
- New disk usage pattern + other small changes
- Add warm start measurement
- Rename install-localhost role to setup-localhost for clarity
- Remove microshift state, unused local cluster state tracking
- Move into ansible subdir
- Add something to README.md

**Which issue(s) this PR addresses**:
This PR adds automation to measure the footprint and performance metrics of MicroShift via Ansible scripts.

### CODE REVIEW

**Comment by dhellmann** [comment]
Let's link this to the Jira ticket

/retitle PERFSCALE-1818: Initial commit of ansible-microshift automation

File: ansible/README.md
Comment (ggiguash) [inline ansible/README.md:4]: Should we replace "as well" by a comma in the sentence?

File: ansible/README.md
Comment (ggiguash) [inline ansible/README.md:13]: Need instructions on installing ansible-galaxy as it's not available in standard repos.

File: ansible/inventory/inventory
Comment (ggiguash) [inline ansible/inventory/inventory:5]: Can we use `microshift` user here?

File: ansible/inventory/inventory
Comment (ggiguash) [inline ansible/inventory/inventory:14]: Can we use `microshift` user here?

File: ansible/roles/epel-repo-install/tasks/main.yml
Comment (ggiguash) [inline ansible/roles/epel-repo-install/tasks/main.yml:19]: We do not support RHEL 9 now, so can we remove this?

File: ansible/roles/install-logging/tasks/main.yml
Comment (ggiguash) [inline ansible/roles/install-logging/tasks/main.yml:10]: We do not support RHEL 9, so this is better removed, I think

File: ansible/roles/install-microshift/defaults/main.yml
Comment (ggiguash) [inline ansible/roles/install-microshift/defaults/main.yml:4]: Why do we need RHEL 9 repos for MicroShift? 
Should we just build RPMs here and install them?
See https://github.com/openshift/microshift/blob/main/docs/devenv_rhel8.md#build-microshift

File: ansible/roles/install-microshift/defaults/main.yml
Comment (ggiguash) [inline ansible/roles/install-microshift/defaults/main.yml:28]: We need firewall configuration for MicroShift as described in https://github.com/openshift/microshift/blob/main/docs/howto_firewall.md. Otherwise, not all pods are going to boot.

File: ansible/roles/install-microshift/tasks/main.yml
Comment (ggiguash) [inline ansible/roles/install-microshift/tasks/main.yml:134]: I think we may want to either build RPMs locally, or download packages from brew.

- https://github.com/openshift/microshift/blob/main/docs/devenv_rhel8.md#build-microshift
- https://github.com/openshift/microshift/blob/main/docs/rpm_packages.md

File: ansible/roles/install-microshift/tasks/main.yml
Comment (ggiguash) [inline ansible/roles/install-microshift/tasks/main.yml:193]: Instead of counting Running pods, we may want to "grep -v Running" and wait for 0. This should be more robust when pods are added.

File: ansible/roles/install-microshift/templates/kubernetes.repo.j2
Comment (ggiguash) [inline ansible/roles/install-microshift/templates/kubernetes.repo.j2:3]: Should it be "el-8"?

File: ansible/inventory/inventory
Comment (dhellmann) [reply ansible/inventory/inventory:5]: Does the user depend on the image?

File: ansible/roles/install-microshift/tasks/main.yml
Comment (dhellmann) [reply ansible/roles/install-microshift/tasks/main.yml:193]: If microshift had a readiness check API endpoint, would we count those core pods as part of the answer it should return?

File: ansible/inventory/inventory
Comment (zshi-redhat) [inline ansible/inventory/inventory:2]: I got this error when filled in the ansible_host and private_ip:
```
fatal: [localhost]: FAILED! => {"msg": "[\"{{ hostvars['microshift'].ansible_host }}\", \"{{ hostvars['logging'].ansible_host }}\"]: \"hostvars['microshift']\" is undefined"}
```

```
microshift.example.com ansible_host=192.168.122.10 private_ip=192.168.122.10
```
How the host vars shall be populated?

File: ansible/roles/epel-repo-install/tasks/main.yml
Comment (zshi-redhat) [reply ansible/roles/epel-repo-install/tasks/main.yml:19]: This is skipped on rhel9:
```
TASK [install-logging : enable powertools repo (CentOS 9)] ******************************************************************************************************************
skipping: [logging.example.com] => {"changed": false, "skip_reason": "Conditional result was False"}
```

File: ansible/inventory/inventory
Comment (sjug) [reply ansible/inventory/inventory:2]: Sorry that was a bit hardcoded, my hosts did not have a hostname so it was looking for a host called `microshift`.
I've pushed a commit to dynamically build the host IPs based on the pattern specified in the inventory here.

File: ansible/README.md
Comment (sjug) [reply ansible/README.md:16]: Added Ansible installation doc step for Fedora.

File: ansible/inventory/inventory
Comment (sjug) [reply ansible/inventory/inventory:5]: Yeah it does, the AMIs for CentOS have user `centos`. It is a var after all so we can set it as necessary.
Once I move to image builder AMIs we can make the default user `microshift`.

File: ansible/inventory/inventory
Comment (sjug) [reply ansible/inventory/inventory:14]: Same as my previous comment for the `microshift:vars` for `ansible_user`.

File: ansible/roles/epel-repo-install/tasks/main.yml
Comment (sjug) [reply ansible/roles/epel-repo-install/tasks/main.yml:19]: It would be skipped on RHEL9 as it's a task for CentOS hosts.
I will change the comment so I don't mention RHEL.

File: ansible/roles/install-logging/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-logging/tasks/main.yml:10]: Would you like me to remove the "RHEL" text from the comment or all the CentOS 9 tasks?

File: ansible/roles/install-microshift/defaults/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/defaults/main.yml:4]: I have created these scripts based upon installation of MicroShift from the COPR repo. 
Though even with building the RPMs, would the build not need prerequisite packages to be installed from the repos?

File: ansible/roles/install-microshift/templates/kubernetes.repo.j2
Comment (sjug) [reply ansible/roles/install-microshift/templates/kubernetes.repo.j2:3]: It should, but there's no `el8` repo.

File: ansible/README.md
Comment (sjug) [reply ansible/README.md]: Yeah that was a pretty bad sentence, let me know if it reads better now.

File: ansible/roles/epel-repo-install/tasks/main.yml
Comment (ggiguash) [reply ansible/roles/epel-repo-install/tasks/main.yml:19]: Yes, I think let's remove this. Otherwise, it's confusing.

File: ansible/roles/install-logging/tasks/main.yml
Comment (ggiguash) [reply ansible/roles/install-logging/tasks/main.yml:10]: Yes, I think let's remove this. Otherwise, it's confusing.

File: ansible/roles/install-microshift/defaults/main.yml
Comment (ggiguash) [reply ansible/roles/install-microshift/defaults/main.yml:4]: COPR packages are outdated. We should probably check it it for the first pass and then discuss our next steps. We need support for RHEL8 as all the development and runtime is now on this OS.

**Review by ggiguash — APPROVED** [review]

File: ansible/roles/install-microshift/defaults/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/defaults/main.yml:31]: Yes sir, updated firewall ports and CIDRs.

File: ansible/roles/install-microshift/defaults/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/defaults/main.yml:4]: Agreed

File: ansible/roles/install-logging/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-logging/tasks/main.yml]: Removed

File: ansible/roles/epel-repo-install/tasks/main.yml
Comment (sjug) [reply ansible/roles/epel-repo-install/tasks/main.yml]: Removed

File: ansible/roles/install-microshift/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/tasks/main.yml]: I've changed the check as you've suggested @ggiguash.
@dhellmann I'm not sure if you're asking me or Gregory, but I assume a readiness check would count the core pods?

Is there any chance of a race where some pods are not yet created but the first ones are in Running state?

File: ansible/roles/install-microshift/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/tasks/main.yml:135]: That's an interesting problem as on one hand we're doing a build on super underpowered hardware (local on the microshift node). Alternatively brew is behind the VPN so not easy to pull to AWS or outside in general.

However we could either build the RPMs locally and then push them to the MicroShift node, or follow that same pattern to pull the RPMs from brew locally and then push them to the node.

File: ansible/roles/install-microshift/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/tasks/main.yml]: The change broke the check, as the query returns "No resources found" for some time until some pods are created.
Any suggestions on what it should be if we don't like the pod running count check that I had before?

Comment (ggiguash) [reply ansible/roles/install-microshift/tasks/main.yml]: Since this is a shell command we're running, we should be able to add a "sleep 1" loop on "No resource found" condition and then proceeding to the next stage once we have at least one pod.

File: ansible/roles/install-microshift/tasks/main.yml
Comment (ggiguash) [reply ansible/roles/install-microshift/tasks/main.yml:135]: MicroShift binary by itself does not require much resources to be built, especially on development VMs.
Anyway, let's push this task to the second round of enhancements.

**Comment by ggiguash** [comment]
Let's merge the first version of this change. I'm approving it and putting it on hold until @sjug feels convenient to release and merge it. Once merged, we should then discuss our next steps.
/hold 
/lgtm

File: ansible/roles/install-microshift/defaults/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/defaults/main.yml]: I've added support for building and installing the MicroShift RPMs from source on RHEL8.

File: ansible/roles/install-microshift/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/tasks/main.yml]: I've added the RPM building on the MicroShift node, takes around 9-10m.

File: ansible/roles/install-microshift/tasks/main.yml
Comment (sjug) [reply ansible/roles/install-microshift/tasks/main.yml:229]: Not really a fan of the current wait situation, but we can discuss this further for future changes.

**Comment by sjug** [comment]
I've addressed pretty much all the review comments and we're in good shape for the most part.
Do you want me to squash and push one commit or do you squash merge?

**Pull request merged**
